radixdlt/src/main/java/com/radixdlt/network/addressbook/PeerManager.java

```diff
@@ -230,55 +231,45 @@ private void heartbeatPeers() {
 	}
 
 	private void handleHeartbeatPeersMessage(Peer peer, SystemMessage heartBeatMessage) {
-		//TODO implement HeartBeat handler
+		log.trace("Received SystemMessage from {}", peer);
 	}
 
 	private void handlePeersMessage(Peer peer, PeersMessage peersMessage) {
 		log.trace("Received PeersMessage from {}", peer);
-		if (peer != null) {
-			List<Peer> peers = peersMessage.getPeers();
-			if (peers != null) {
-				EUID localNid = this.localSystem.getNID();
-				peers.stream()
+		List<PeerWithSystem> peers = peersMessage.getPeers();
+		if (peers != null) {
+			EUID localNid = this.localSystem.getNID();
+			List<PeerWithSystem> unknownPeers = peers.stream()
 				.filter(Peer::hasSystem)
-				.filter(p -> !localNid.equals(p.getNID()))
-				.forEachOrdered(addressbook::updatePeer);
+				.filter(p -> !localNid.equals(p.getNID())) // Not us
+				.filter(p -> this.addressbook.peer(p.getNID()).isEmpty()) // Not already in address book
+				.filter(p -> p.supportedTransports().findAny().isPresent()) // Has some transports

```

Comment:
I think that for filters it makes sense to create dedicated methods with appropriate names. This will improve readability and make comments unnecessary.

Reply:
OK, methods added.